### DISCUSSION TITLE
Editor needs to gracefully handle files being renamed or moved outside of the editor

### DISCUSSION THREAD

**Discussion by blackears** [godotengine/godot-proposals#6907] [Editor]
I've been developing an addon for Godot and am running into huge problems whenever I try to move files outside of the Godot editor.  If I simply move files to a different directory or rename them and then try to open the project in Godot again, I get a ton of error messages about the files either not being findable or being duplicates of other files.  This is effectively preventing me from using branching and merging with git.  Even just moving a group of files into a new folder for organization causes my project to be unusable.

If I copy and paste the code into a new project and load that, it works fine.  But my old repository has a ton of errors and I can't figure out a way to just tell the editor to forget everything it has cached and start new.  One person suggested disabling and reenabling the plugin, but that didn't fix anything.  In the short term I need a work around so that I can use my repository again.

Attached is a simple project where I created an addon and also created a file named Foo.gd in the root directory.  I gave Foo.gd the `class_name` Foo and create a new instance of it in the addon's main object.  At this point the addon was working, even if I closed and reloaded Godot.  Then I closed Godot and created a subdirectory called `subdir` and moved Foo.gd into it using Windows Explorer.  When I next loaded the project in Godot I got the errors

`
res://addons/testplugin/testplugin.gd:4 - Parse Error: Could not parse global class "Foo" from "res://addons/testplugin/Foo.gd".
  res://addons/testplugin/testplugin.gd:7 - Parse Error: Could not find script for class "Foo".
  Attempt to open script 'res://addons/testplugin/Foo.gd' resulted in error 'File not found'.
  res://addons/testplugin/testplugin.gd:-1 - Compile Error:
`

While I think this is a bug that needs to be fixed, in the short term I just need a workaround so I can make my repository usable again.  Is there  a way to do this?

[AddonMoveFiles.zip](https://github.com/godotengine/godot-proposals/files/11504378/AddonMoveFiles.zip)

**Comment by theraot** [comment]
Note that the attached zip does not include the `.godot` folder, as a result I get the error messages on the first load, but Godot regenerates the `.godot` folder and works correctly, despite having shown the error messages.

There were no subsequent error messages reloading the project, disabling and enabling the plugin, nor adding an scene and running it. Furthermore, I tried moving the file around both while Godot was running and while it wasn't, and Godot got the changes correctly.

I also tried introducing a parse error in the file, closing Godot, moving the file, and opening Godot again (which failed to enable the plugin, of course), and fixing the parse error in Godot, and enabling the plugin. Despite the spill of error messages it worked correctly.

I have been unable to find a situation where the error messages don't go away. So, I believe Godot is being eager in showing these error messages despite being able to recover correctly.

---

Anyway, I suppose there might be a situation where forcing to regenerate the cache is necessary. I just have been unable to find it. So know that the cache for the list of classes is in `.godot/global_script_class_cache.cfg` and you could edit with any text editor if it comes to it... Or remove it and have Godot regenerate it. That might be the solution for your old project.

**Comment by theraot** [comment]
OK, I managed to get the issue to happen on one of my projects. In this case I moved a script that is referenced in a scene (attached to one of the nodes). *And, you know, I thought ids would fix that.* Apparently Godot saved the script again creating a duplicate. Removing the script made it work, but turns out Godot saved the script inside the scene file instead of looking for the file elsewhere. *Again, I would have expected ids would fix that.*

Addendum: The solution was to attach the correct script.

**Comment by theraot** [comment]
I have become convinced ids are just not reliable enough for this task. This is how:

I have a scene with a child node with a script attached. Saved. I take note of the id Godot generated. I remove the node and reload the project. Then I add a new node with the same script, and take note of the id that Godot generated. The ids are different. And this is the same machine. *In fact, I didn't find a persisted cache of id to path. Adding such cache would be one more thing that can go out of sync anyway.*

So in the situation where we are merging changes made in different machines at different times, we cannot trust ids to match.

Because of this I would propose to **use the class name as fallback when available**. This means adding the class name to the scene files where the script is referenced, and when the script is not found at the path saved in the scene file look for a script matching the class name and use that - yes, I think this should take precedence over the id.

**Comment by blackears** [comment]
Here is the project with the .godot file.  I left it out because your submission process for regular issues requests that you do so.

[AddonMoveFiles.zip](https://github.com/godotengine/godot-proposals/files/11509169/AddonMoveFiles.zip)

**Reply by theraot** [reply]
**About the project in the original post**

I only got error messages from the project in the original post. But these don't *persist*. I mean, despite getting error messages from the project in the original post, the code works without issue and the error messages don't come back.

---
**About this and the prior projects you shared**

This, nor the other one (https://github.com/godotengine/godot-proposals/discussions/6907#discussioncomment-5940146) give me any issues. I did not even get error messages on first opening in either of these two.

---
**About the other issue I found**

The issue I found (https://github.com/godotengine/godot-proposals/discussions/6907#discussioncomment-5938653) was having the script to a node in a scene, which you don't have in the projects you share.

---

So, does the error *persist* despite not using the script in a scene? - That would suggest I have not been able to reproduce your particular issue, and the issue I found is a different one.

If the error is related to a scene - as is the case the other issue I found - then the solution would be to fix the reference in the scene.

---

*I'm testing on Windows, by the way.*

**Reply by blackears** [reply]
The code is valid so I would expect it to work on your machine.  The problem is moving files outside of the Godot project.  So if you were to close the project after opening it, move Foo.gd to a different directory and then reopen the project, I would expect you would get errors like I'm getting.  

Also, since you mentioned the .godot files as containing the cache, I included them in case you needed to examine them specifically.  That is the only change.  Deleting the .godot file and restarting the editor does not fix my problem.

**Reply by blackears** [reply]
The problem is not the code itself since I can copy and paste it into a different project and it works fine.  The problem is that moving/renaming files outside of Godot and then reopening the project causes Godot to not be able to find them.

**Reply by blackears** [reply]
Since this is an addon, you also need to make sure the addon is enabled.   I've added the @tool tag to Foo.gd since that ought to be there.  So here is an updated process that breaks things on my system.

* open the addon.  Check the Output window to make sure there are no errors.
* Make sure the addon is enabled in the plugins Window.
* Close Godot
* Move Foo.gd from subdir to subdir2
* Open the Godot editor and select the AddonMoveFiles project
* Output window will now have error messages like

`  res://addons/testplugin/testplugin.gd:4 - Parse Error: Could not parse global class "Foo" from "res://addons/testplugin/subdir2/Foo.gd".
  res://addons/testplugin/testplugin.gd:7 - Parse Error: Could not find script for class "Foo".
  Attempt to open script 'res://addons/testplugin/subdir2/Foo.gd' resulted in error 'File not found'.
  res://addons/testplugin/testplugin.gd:-1 - Compile Error:`

[AddonMoveFiles.zip](https://github.com/godotengine/godot-proposals/files/11510970/AddonMoveFiles.zip)

**Reply by theraot** [reply]
> The problem is moving files outside of the Godot project.

I also mentioned (here: https://github.com/godotengine/godot-proposals/discussions/6907#discussioncomment-5937705) that I tried all that with the project in the original post.

I reiterate that the error messages I got don't persist. They only show once, and the project works. I had to do **nothing**. If that is your case, then we are only arguing if Godot is being too eager to show the error messages, despite everything being fine. Yet, I suspect that is not the case for you since you are asking for a fix or workaround.

So you move the files outside of Godot, Godot shows the error messages, and then what? Does the project work correctly? Does it keep showing the errors every time? Does reloading the project fix it? I suspect that in your case the project does not work, keeps showing the error message, and reloading does not fix it (I don't know, I'm guessing)...

Once more, I got a situation which involves a scene where the error does not go away. Is that your case? I guessing it isn't because you haven't included an scene in the projects you shared. Which would mean I found a different problem.

So, what else does it take for me to get the same situation as you have on my end? *I did suspect some editor setting could be the reason I don't get the same problem as you, but I can't tell which if any.*

**Reply by blackears** [reply]
Actually I just tried reopening the project for  a second time and didn't get an error  that time.  My real issue is that I'm trying to work on https://github.com/blackears/cyclopsLevelBuilder.  Two days ago I tried branching the project so that I could make some big changes and started by moving most of the files under /actions, /commands/ and /tools into new directories.  Now the plugin no longer runs properly (it should show a scene in the viewport) and every time I open it I get a ton of errors.  Reverting everything did not fix this problem.  

I've also been plagued by problems like this all though the development process.  Renaming or moving a file causes errors to show when I next open the project.  I've noticed that closing and reopening Godot twice often caused the error to go away, but after moving so many files my repo seems to be permanently messed up.  Or maybe Godot is just fixing one load error per restart.

**Reply by blackears** [reply]
Restarting multiple times does not fix the errors I get.  The code itself is fine since I can copy and paste it into a new project and it runs.  Deleting the .godot file and having the editor rebulid it does not fix the problem.  I may have made some other changes too, but after reverting everything and switching back to the main branch, I would have expected to at least be able to run again.  But I still get errors no matter how many times I restart.

**Reply by theraot** [reply]
@blackears I got a copy from https://github.com/blackears/cyclopsLevelBuilder and opened it in Godot, it gave me a bunch of errors. Disabling the plugin and enabling it again didn't make them go away. Godot kept outputting error messages and increasing memory usage until I had to kill the process. As far as I can tell this because some editor scripts were still running. Anyway... After loading Godot again, it was fixed.

I ran a diff to see what changed, and the only thing was that Godot made the .godot folder (and a minor change in the scene caused by the editor scripts, but reverting that didn't bring the errors back). What did bring the errors back was removing the .godot folder. So for the second time I closed Godot normally, and opened it again. And once more the errors were gone.

**Reply by theraot** [reply]
@blackears I just tried moving one of the files in the addon. And, yes Godot gave me a bunch of error upon reopening it. So, I tried closing it and opening it again. This time I got only one unrelated error. Closed and opened again. And it opened with no error.

However trying to use the plugin gave me an error.

For full context, I move the file `numeric_line_edit.gd`, and the error I got when trying to use the plugin was `res://addons/cyclops_level_builder/docks/uv_editor/uv_editor_viewport.gd:61 - Trying to assign value of type 'numeric_line_edit.gd' to a variable of type 'numeric_line_edit.gd'.`

The script `uv_editor_viewport.gd` was attached to a node in a scene. The scene was still pointing to the old path, but Godot didn't re-create the file... Instead trying to open the script from the scene opens like an empty file, plus trying to show it in the filesystem results in an error.

I'm not sure how it ended up like that, but I believe my idea of using the class name to resolve it might address it.

**Reply by blackears** [reply]
I've been getting a lot of strange behavior.  I only have my development machine and my laptop to work with, so it's hard to tell what others are getting.  I'll be developing for a few days and things are going fine and I'm happy with the code, but then I'll make a small change and then the whole project breaks.  After a lot of saving, reloading and reverting, I finally get things back to a working state.  The only thing that I can think of to explain it is Godot is somehow completely messing up class names or file locations.

A few days ago I tried checking the project out on my Macbook Pro laptop for the first time, and I get no end of error messages on it.  Reloading the project several times does not help.

On my dev machine, I did a fresh clone of the repository and while it did give me errors the first time I loaded it, when I loaded it the second time it worked, so I'm developing on that now.  I can't use my first clone no matter how many times I open it.

I am worried about what my users are seeing.  I know some of them are having it run because I'm getting feedback that indicates they can run it.  But how many of them download a copy, run into a wall of error messages and never try it again?

I can't dig into the code (my machine doesn't support Vulkan and the last time I tried to build the dev branch of Godot it would not run without Vulkan) and have little idea of that is going on in the editor.  But something is going really wrong with how Godot is keeping track of class names and where files are located.

**Reply by blackears** [reply]
@theraot I'm not quite sure from your post how you used a class name to resolve an issue.  I've been naming all my classes and even use class inheritance in a few places - I have a background in C++ and C# and like to use those features.

The `uv_editor_viewport.gd` is a control and not one of the files I was trying to move.  I can see how moving that would have caused a problem since it is referenced by a scene node.  Most of the files I moved are not referenced that way. And in any case, after I reverted everything they should have been back in their old spot.  (The files I was moving were in /actions, /commands and /tools which provide the functionality of the addon).  Some were deleted instead of moved.  While I would have expected some errors, my first repo is now completely locked by them no matter how many times I reload.  And deleting the .godot folder does not change anything.

**Reply by theraot** [reply]
@blackears My idea is to modify the way Godot saves the references to the scripts in the resource files (including scene files) to include the class name, and when the file is not found try to find it by class name (as I was saying here: https://github.com/godotengine/godot-proposals/discussions/6907#discussioncomment-5938930). Godot does not currently do what I'm saying. I need to write either an issue or proposal, probably both, and then make a pull request. I'm looking at existing related issues and proposals first.

Looking at the files you mention, the ones on /actions and /commands appear to be only referenced by class name, but there are files in /tools that are referenced in resource files (the ones in /data/tool_tags). Regenerating the .godot folder won't fix those. I'm guessing these are the issue, and also that the change I have in mind would address them too.

**Reply by theraot** [reply]
@blackears Anyway, for your fix before any changes to Godot, you would have to find the problematic references and fix them.

**Reply by blackears** [reply]
@theraot  There aren't any problematic references because everything has been reverted to it's original state.  All the references should be valid (though I still get errors when I load).  But your proposal does sound like a good one and I hope it fixes the issue.

**Reply by theraot** [reply]
@blackears Well, I don't know. I don't see anything wrong with the paths. They are using "/", and fully in lower case. No special characters, or names that would work on one OS but not the other. Plus you say you have reverted the changes and removing the .godot folder does not fix it. So no idea.

Anyway, I found that "Edit dependencies..." -> "Fix Broken" fixes the problem I was describing. Which also means the issue is that Godot is not prompting the user to do that. So I opened an issue about it.

**Comment by blackears** [comment]
I've tried deleting the .godot file before, but that does not solve the problem for me.  Perhaps this is a Windows specific issue?  Anyhow, in this run I renamed .godot to .godot_ so the editor wouldn't find it and reloaded my scene.  I'm still getting errors:

![image](https://github.com/godotengine/godot-proposals/assets/6724608/6b7f7e38-4c2f-4446-8455-6efcf6e3abab)

[AddonMoveFiles.zip](https://github.com/godotengine/godot-proposals/files/11509274/AddonMoveFiles.zip)
